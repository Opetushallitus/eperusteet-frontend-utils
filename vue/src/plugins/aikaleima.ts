import Vue from 'vue';
import moment from 'moment';

export class Aikaleima {
  public install(vue: typeof Vue) {
    function aikaleimaFnFactory(this: void, format: string) {
      // eslint-disable-next-line only-arrow-functions
      return function(this: void, value: number) {
        (this as any).$i18n.locale; // eslint-disable-line
        return moment(value).format(format);
      };
    }

    // Long datetime
    vue.prototype.$ldt = aikaleimaFnFactory('LLL');

    // Long date
    vue.prototype.$ld = aikaleimaFnFactory('LL');

    // Long time
    vue.prototype.$lt = aikaleimaFnFactory('H:mm:ss');

    // Short datetime
    vue.prototype.$sdt = aikaleimaFnFactory('D.M.YYYY H:mm');

    // Short date
    vue.prototype.$sd = aikaleimaFnFactory('D.M.YYYY');

    // Short time
    vue.prototype.$st = aikaleimaFnFactory('H:mm');

    // Time until or ago an event counting from now
    vue.prototype.$ago = function(value: number) {
      this.$i18n.locale; // eslint-disable-line
      return moment(value).fromNow();
    };

    // Custom datetime
    vue.prototype.$cdt = function(value: number, format: string) {
      // Pakko olla, jotta localen vaihtuessa komponentti p√§ivittyy
      this.$i18n.locale; // eslint-disable-line
      return moment(value).format(format);
    };
  }
}

export default new Aikaleima();
