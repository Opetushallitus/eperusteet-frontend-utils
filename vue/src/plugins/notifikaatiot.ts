import Vue from 'vue';

interface NotificationConfig {
  title: string;
  kind?: 'info' | 'warn' | 'error' | 'success';
  text?: string;
}

export interface CheckedConfig {
  success?: string;
  failure?: string;
}

declare module 'vue/types/vue' {
  interface Vue {
    $notification: (config: NotificationConfig) => Promise<void>;
    $success: (title: string) => Promise<void>;
    $info: (title: string) => Promise<void>;
    $fail: (title: string, text?: string) => Promise<void>;
    $warning: (title: string, text?: string) => Promise<void>;
  }
}

export class Notifikaatiot {
  public static install(vue: typeof Vue) {
    if (!vue.prototype.$notify) {
      throw new Error('Vue.use(require("vue-notification"))');
    }

    vue.prototype.$notification = function(config: NotificationConfig) {
      this.$notify({
        title: config.title,
        type: config.kind || 'info',
        text: config.text,
      });
    };

    vue.prototype.$success = function(title: string) {
      this.$notify({
        title,
        type: 'success',
      });
    };

    vue.prototype.$info = function(title: string) {
      this.$notify({
        title,
        type: 'info',
      });
    };

    vue.prototype.$fail = function(title: string, text: string = '') {
      this.$notify({
        title,
        type: 'error',
        text,
        duration: 5000,
      });
    };

    vue.prototype.$warning = function(title: string, text: string = '') {
      this.$notify({
        title,
        type: 'warn',
        text,
        duration: 5000,
      });
    };
  }
}
