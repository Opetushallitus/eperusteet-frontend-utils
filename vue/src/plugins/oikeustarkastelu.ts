import Vue from 'vue';
import _ from 'lodash';
import { Oikeus } from '../tyypit';
import { Computed } from '../utils/interfaces';

const DisableTags = ['input', 'button'];

export interface IOikeusProvider {
  hasOikeus: (oikeus: Oikeus, kohde?: any) => boolean;
  isAdmin?: Computed<boolean>;
  hasOphCrud?: Computed<boolean>;
  casKayttaja: Computed<any>;
  sovellusOikeudet: Computed<SovellusOikeus[]>;
  logoutUrl?: Computed<string>;
}

export interface OikeustarkasteluConfig {
  oikeusProvider: IOikeusProvider;
}

declare module 'vue/types/vue' {
  interface Vue {
    $hasOikeus: (oikeus: Oikeus, kohde?: any) => boolean;
    $isAdmin: () => boolean;
    $hasOphCrud: () => boolean;
  }
}

export class Oikeustarkastelu {
  public static install(vue: typeof Vue, config: OikeustarkasteluConfig) {
    vue.prototype.$isAdmin = () => config.oikeusProvider.isAdmin?.value || false;
    vue.prototype.$hasOphCrud = () => config.oikeusProvider.hasOphCrud?.value || false;

    vue.prototype.$hasOikeus = function(oikeus: Oikeus, kohde?: any) {
      return config.oikeusProvider.hasOikeus(oikeus, kohde);
    };

    // Sisällön kääntäminen
    vue.directive('oikeustarkastelu', (el: any, binding) => {
      // Hide the element before rights have been resolved
      if (!el.oldDisplayValue && el.style.display !== 'none') {
        el.oldDisplayValue = el.style.display;
      }

      el.style.display = 'none';

      let value = binding.value || 'luku';
      let kohde: any;
      if (_.isObject(value)) {
        kohde = (value as any).kohde;
        value = (value as any).oikeus;
      }

      if (config.oikeusProvider.hasOikeus(value, kohde)) {
        el.style.display = el.oldDisplayValue;
      }
      else {
        const { tagName } = el;
        if (_.includes(DisableTags, _.toLower(tagName))) {
          (el as HTMLInputElement).disabled = true;
          el.style.display = el.oldDisplayValue;
        }
      }
    });
  }
}

export interface SovellusOikeus {
  eperusteSovellus: EperusteSovellus;
  valittu: boolean;
}

export interface EperusteSovellus {
  sovellus: string;
  url: string;
}

export const EPERUSTEET_SOVELLUKSET = [
  {
    sovellus: 'APP_EPERUSTEET',
    url: '/eperusteet-app',
  },
  {
    sovellus: 'APP_EPERUSTEET_YLOPS',
    url: '/eperusteet-ylops-app',
  },
  {
    sovellus: 'APP_EPERUSTEET_AMOSAA',
    url: '/eperusteet-amosaa-app/#/ammatillinen',
  },
  {
    sovellus: 'APP_EPERUSTEET_VST',
    url: '/eperusteet-amosaa-app/#/vapaasivistystyo',
  },
  {
    sovellus: 'APP_EPERUSTEET_TUVA',
    url: '/eperusteet-amosaa-app/#/tutkintoonvalmentava',
  },
  {
    sovellus: 'APP_EPERUSTEET_KOTO',
    url: '/eperusteet-amosaa-app/#/kotoutumiskoulutus',
  },
];

export function getSovellusoikeudet(casRoles: string[], valittuSovellus: string): SovellusOikeus[] {
  const toSovellusOikeus = (eperusteSovellus) => ({
    eperusteSovellus,
    valittu: valittuSovellus === eperusteSovellus.sovellus,
  });

  const origin = window.location.origin;
  if (_.includes(origin, 'localhost')) {
    return _.map(EPERUSTEET_SOVELLUKSET, eperusteSovellus => toSovellusOikeus(eperusteSovellus));
  }
  else {
    return _.chain(casRoles)
      .filter(casRole => _.includes(_.map(EPERUSTEET_SOVELLUKSET, 'sovellus'), casRole))
      .map(sovellus => _.find(EPERUSTEET_SOVELLUKSET, eSovellus => eSovellus.sovellus === sovellus))
      .map(sovellus => toSovellusOikeus(sovellus))
      .value();
  }
}

export const OIKEUS_KAANNOT = {
  'READ': 'luku',
  'READ_UPDATE': 'muokkaus',
  'CRUD': 'luonti',
  'ADMIN': 'hallinta',
};

export const EPERUSTEET_KOULUTUSTYYPPI_PAIKALLISET_SOVELLUKSET = {
  'koulutustyyppi_1': 'APP_EPERUSTEET_AMOSAA', // ammatillinen perustutkinto
  'koulutustyyppi_2': 'APP_EPERUSTEET_YLOPS', // lukiokoulutus
  'koulutustyyppi_5': 'APP_EPERUSTEET_AMOSAA', // telma
  'koulutustyyppi_6': 'APP_EPERUSTEET_YLOPS', // lisaopetus
  'koulutustyyppi_11': 'APP_EPERUSTEET_AMOSAA', // ammattitutkinto
  'koulutustyyppi_12': 'APP_EPERUSTEET_AMOSAA', // erikoisammattitutkinto
  'koulutustyyppi_14': 'APP_EPERUSTEET_YLOPS', // aikuisten lukiokoulutus
  'koulutustyyppi_15': 'APP_EPERUSTEET_YLOPS', // esiopetus
  'koulutustyyppi_16': 'APP_EPERUSTEET_YLOPS', // perusopetus
  'koulutustyyppi_17': 'APP_EPERUSTEET_YLOPS', // aikuisten perusopetus
  'koulutustyyppi_18': 'APP_EPERUSTEET_AMOSAA', // valma
  'koulutustyyppi_20': 'APP_EPERUSTEET_YLOPS', // varhaiskasvatus
  'koulutustyyppi_22': 'APP_EPERUSTEET_YLOPS', // perusopetuksen valmistava
  'koulutustyyppi_23': 'APP_EPERUSTEET_YLOPS', // valmistava lukiokoulutus
  'koulutustyyppi_999907': 'APP_EPERUSTEET_YLOPS', // taiteen perusopetus,
  'koulutustyyppi_10': 'APP_EPERUSTEET_VST', // vapaa sivistystyo
  'koulutustyyppi_30': 'APP_EPERUSTEET_KOTO', // aikuisten maahanmuuttajien kotoutumiskoulutus
  'koulutustyyppi_35': 'APP_EPERUSTEET_VST', // vapaa sivistystyo lukutaito
  'koulutustyyppi_40': 'APP_EPERUSTEET_TUVA', // tutkintoon valmentava
};
