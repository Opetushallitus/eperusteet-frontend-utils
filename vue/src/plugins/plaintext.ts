import Vue from 'vue';

export class Plaintext {
  public install(vue: typeof Vue) {
    vue.prototype.$plaintext = (str: string) => str.replace(/<\/?[^>]+(>|$)/g, '');
  }
};

export default new Plaintext();
