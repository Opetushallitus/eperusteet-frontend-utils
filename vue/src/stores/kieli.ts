import { createLogger } from '../utils/logger';
import { Kieli } from '../tyypit';
import _ from 'lodash';
import VueI18n from 'vue-i18n';
import Vue, { VueConstructor } from 'vue';
import moment from 'moment';
import 'moment/locale/en-gb';
import 'moment/locale/fi';
import 'moment/locale/ru';
import 'moment/locale/se';
import 'moment/locale/sv';
import VueCompositionApi, { computed, reactive } from '@vue/composition-api';
import { updateRelativeTime } from '../plugins/aikaleima';
import { Computed } from '../utils/interfaces';

Vue.use(VueCompositionApi);

declare module 'vue/types/vue' {
  interface Vue {
    $locale: Computed<string>;
    $slang: Computed<string>;
    $t: typeof VueI18n.prototype.t;
    $tc: typeof VueI18n.prototype.tc;
    $te: typeof VueI18n.prototype.te;
  }
}

export const UiKielet = Object.freeze(['fi', 'sv', 'en']);
const logger = createLogger('Kieli');

function getMessages() {
  const result: any = {};
  try {
    result.fi = require('@/translations/locale-fi.json');
  }
  catch (e) {}

  try {
    result.sv = require('@/translations/locale-sv.json');
  }
  catch (e) {}
  return result;
}

export class KieliStore {
  private vi18n!: VueI18n;

  public get i18n() {
    return this.vi18n;
  }

  /**
   * Add language support to root vue instance.
   *
   */
  public install(v: VueConstructor, config = {} as Partial<VueI18n.I18nOptions>) {
    moment.locale(Kieli.fi);
    this.vi18n = new VueI18n({
      fallbackLocale: Kieli.fi,
      locale: Kieli.fi,
      messages: getMessages(),
      ..._.cloneDeep(config),
    });
    v.prototype.$locale = this.uiKieli;
    v.prototype.$slang = this.sisaltoKieli;
    moment.updateLocale(Kieli.fi, updateRelativeTime());
  }

  /**
   * Load translation from other source
   *
   */
  public async load(kaannokset: Kaannokset) {
    logger.info('Initing locales');
    const results = await this.fetchLocaleMap(kaannokset);
    _.forEach(results, (locales, lang) => {
      this.i18n!.mergeLocaleMessage(lang, locales);
    });
  }

  private readonly state = reactive({
    sisaltoKieli: Kieli.fi,
  });

  public readonly getUiKieli = computed(() => this.i18n?.locale);
  public readonly getSisaltoKieli = computed(() => this.state.sisaltoKieli);
  public readonly sisaltoKieli = computed(() => this.state.sisaltoKieli);
  public readonly uiKieli = computed(() => this.i18n?.locale);

  public readonly getAikakaannokset = computed(() => {
    const kieli = this.state.sisaltoKieli;
    return {
      days: _.map(moment.weekdays(), (day: string) => _.toUpper(_.first(day))),
      months: moment.monthsShort(),
      placeholder: {
        date: this.i18n!.t('valitse-pvm'),
        dateRange: this.i18n!.t('valitse-pvm-jana'),
      },
    };
  })

  public setUiKieli(kieli: Kieli) {
    if (this.i18n!.locale !== kieli && _.includes(UiKielet, kieli)) {
      moment.locale(kieli);
      this.i18n!.locale = kieli;
    }
  }

  public setSisaltoKieli(kieli: Kieli) {
    if (this.state.sisaltoKieli !== kieli && _.includes(UiKielet, kieli)) {
      this.state.sisaltoKieli = kieli;
    }
  }

  public haeLokalisoituOlio(avain: string) {
    const result = {
      fi: this.i18n!.t(avain, 'fi'),
      sv: this.i18n!.t(avain, 'sv'),
      en: this.i18n!.t(avain, 'en'),
    };
    return result;
  }

  /**
   * Check if text query matches object
   *
   */
  public search(query: string, text: any, config = {}): boolean {
    if (text && query) {
      const target = _.isString(text) ? text : text[this.getSisaltoKieli.value];
      return _.includes(_.toLower(target), _.toLower(query));
    }
    else {
      return true;
    }
  }

  public filterBy(field: string, query: string, config = {}) {
    return (data: any) => this.search(query, data[field]);
  }

  public searchFn(query: string) {
    return (text: any) => this.search(query, text);
  }

  /**
   * i18n locale convert
   *
   */
  public t(value: string): string {
    if (this.vi18n) {
      return this.vi18n.t(value) as string || '<' + value + '>';
    }
    else {
      return value;
    }
  };

  /**
   * Convert field into sortable value
   *
   * @returns {string}
   */
  public sortValue(value: any): string {
    if (!value) {
      return '';
    }
    else if (_.isObject(value)) {
      const locale = this.getSisaltoKieli.value;
      return (value as any)[locale];
    }
    else {
      return value;
    }
  };

  public kaanna(value: LokalisoituTeksti | undefined | null): string {
    if (!value) {
      return '';
    }
    else if (_.isObject(value)) {
      const locale = this.getSisaltoKieli.value;
      const kielet = [locale, ..._.pull(['fi', 'sv', 'en', 'se', 'ru'], locale)];
      let teksti = '';

      _.forEach(kielet, kieli => {
        if (!_.isEmpty(value[kieli])) {
          teksti = value[kieli] as string;
          return false;
        }
      });

      return teksti;
    }
    else {
      logger.warn('"$kaanna" on tekstiolioiden kääntämiseen. Käytä vue-i18n vastaavaa funktiota. Esimerkiksi "$t()".', 'Käännös:', '"' + value + '"');
      return value;
    }
  };

  public kaannaOlioTaiTeksti(value: LokalisoituTeksti | string): string {
    if (_.isObject(value)) {
      return this.kaanna(value);
    }
    else if (_.isString(value)) {
      return this.t(value);
    }
    else {
      return value;
    }
  }

  private async fetchLocaleMap(kaannokset: Kaannokset) {
    try {
      const result: any = {};
      const localeObj = kaannokset;
      _.forEach(localeObj, (locales, lang) => {
        result[lang] = {};
        if (locales) {
          for (const locale of locales) {
            if (locale.key && locale.value) {
              result[lang][locale.key] = locale.value;
            }
          }
        }
      });
      return result;
    }
    catch (err) {
      logger.error('Käännösten haku epäonnistui', err.message);
      return {};
    }
  }
}

export const Kielet = new KieliStore();

export interface Kaannos {
  key: string;
  value: string;
}

export interface Kaannokset {
  fi?: Kaannos[];
  sv?: Kaannos[];
  en?: Kaannos[];
}

export interface LokalisoituTeksti {
  [locale: string]: string | null | undefined | number;
  _id?: number;
  _tunniste?: string;
}
