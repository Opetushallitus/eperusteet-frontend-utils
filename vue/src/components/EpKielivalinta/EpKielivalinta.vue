<template>
   <b-dropdown size="sm" class="kielivalinta" right>
    <template slot="button-content">
      <span>{{ $t("kieli-sisalto") }}:
        <span class="valittukieli">{{ $t(sisaltoKieli) }}</span>
        <EpMaterialIcon>expand_more</EpMaterialIcon>
      </span>
    </template>
    <b-dropdown-item @click="valitseSisaltoKieli(kieli)" v-for="kieli in sovelluksenKielet" :key="kieli" :disabled="kieli === sisaltoKieli">{{ $t(kieli) }}</b-dropdown-item>
  </b-dropdown>

</template>

<script lang="ts">
import { Component, Vue } from 'vue-property-decorator';
import { Kielet, UiKielet } from '@shared/stores/kieli';
import { Kieli } from '@shared/tyypit';
import EpMaterialIcon from '@shared/components/EpMaterialIcon/EpMaterialIcon.vue';

@Component({
  components: {
    EpMaterialIcon,
  },
})
export default class EpKielivalinta extends Vue {
  get sisaltoKieli() {
    return Kielet.getSisaltoKieli.value;
  }

  get sovelluksenKielet() {
    return UiKielet;
  }

  valitseSisaltoKieli(kieli: Kieli) {
    Kielet.setSisaltoKieli(kieli);
  }
}
</script>

<style scoped lang="scss">
@import "@shared/styles/_variables.scss";

  /deep/ .dropdown-toggle::after {
    display:none;
  }

  /deep/ .btn {
    background-color: $white !important;
    color: $black  !important;
    border-color: $white !important;
    box-shadow: 0;
  }

  /deep/ .btn-secondary,
  /deep/ .btn-secondary:not(:disabled):not(.disabled):active:focus,
  /deep/ .btn-secondary:not(:disabled):not(.disabled).active:focus,
  /deep/ .show > .btn-secondary.dropdown-toggle:focus {
    box-shadow: none;
  }

  .valittukieli {
    padding: 5px 70px 5px 10px;
    border: 1px solid $gray-lighten-3;
    border-radius: 5px;
  }

  .valintanuoli {
    margin-left: -20px;
  }

</style>
