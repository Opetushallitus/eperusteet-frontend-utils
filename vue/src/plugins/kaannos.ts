import Vue from 'vue';
import { createLogger } from '../utils/logger';
import { Kielet } from '../stores/kieli';
import * as _ from 'lodash';
import { unescapeStringHtml } from '@shared/utils/inputs';

const logger = createLogger('Kaannos');

declare module 'vue/types/vue' {
  interface Vue {
    $suodatin: (query: string) => <T extends object>(value: T) => string;
    $filterBy: (field: string, query: string) => <T extends object>(value: T) => boolean;
    $kaanna: <T extends object>(value?: T, emptyWhenNotFound?: boolean, squareBrackets?: boolean) => string;
    $kaannaOlioTaiTeksti: <T extends object>(value: T | string) => string;
  }
}

export class Kaannos {
  public install(vue: typeof Vue, options) {
    // Sisällön kääntäminen
    vue.prototype.$suodatin = (str: string) => Kielet.searchFn(str);

    vue.prototype.$filterBy = function(field: string, query: string) {
      return Kielet.filterBy(field, query);
    };

    vue.prototype.$kaanna = (value?: any, emptyWhenNotFound = false, squareBrackets = true) => {
      return this.handleUnescaping(Kielet.kaanna(value, emptyWhenNotFound, _.has(options, 'squareBrackets') ? options.squareBrackets : squareBrackets));
    };

    vue.prototype.$kaannaOlioTaiTeksti = (value: any, emptyWhenNotFound = false, squareBrackets = true) => {
      return this.handleUnescaping(Kielet.kaannaOlioTaiTeksti(
        value,
        _.has(options, 'emptyWhenNotFound') ? options.emptyWhenNotFound : emptyWhenNotFound,
        _.has(options, 'squareBrackets') ? options.squareBrackets : squareBrackets));
    };
  }

  // Ei tehdä unescapetusta, jos sisällössä oleva sisäinen linkki sisältää routenode-attribuutin, koska rikkoo tämän json-rakenteen.
  private handleUnescaping(kaannos) {
    return _.includes(kaannos, 'routenode') ? kaannos : unescapeStringHtml(kaannos);
  }
}

export default new Kaannos();
