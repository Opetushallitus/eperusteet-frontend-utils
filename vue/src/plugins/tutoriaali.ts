import { TutoriaaliStore } from '../stores/tutoriaali';
import Vue from 'vue';

interface VueTutorialParams {
  tutoriaaliStore: TutoriaaliStore;
}

export class VueTutorial {
  public static install(vue: typeof Vue, options: VueTutorialParams) {
    if (!options?.tutoriaaliStore) {
      throw new Error('TutoriaaliStore puuttuu');
    }

    vue.directive('tutorial', {
      inserted(el, def) {
        el.setAttribute('tutorial', '');
        options.tutoriaaliStore.paivitaAvaimet();
      },
    });
  }
};
