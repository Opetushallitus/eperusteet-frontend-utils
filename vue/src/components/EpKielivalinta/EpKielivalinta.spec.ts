import { mount, createLocalVue, RouterLinkStub } from '@vue/test-utils';
import EpKielivalinta from './EpKielivalinta.vue';
import { Kaannos } from '../../plugins/kaannos';
import VueI18n from 'vue-i18n';
import { Kielet } from '../../stores/kieli';
import Vue from 'vue';
import BootstrapVue from 'bootstrap-vue';

Vue.use(BootstrapVue);

describe('EpKielivalinta component', () => {
  const localVue = createLocalVue();
  localVue.use(VueI18n);
  Kielet.install(localVue);
  localVue.use(new Kaannos());

  test('Renders', async () => {
    const wrapper = mount(EpKielivalinta, {
      localVue,
      mocks: {
        $t: x => x,
      },
      stubs: {
        'router-link': RouterLinkStub,
      },
    });

    expect(wrapper.html()).toMatchSnapshot();
  });
});
