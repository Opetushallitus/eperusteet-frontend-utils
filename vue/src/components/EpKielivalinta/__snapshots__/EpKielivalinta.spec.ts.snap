// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EpKielivalinta component Renders 1`] = `
<div class="dropdown b-dropdown kielivalinta btn-group">
  <!----><button aria-haspopup="true" aria-expanded="false" type="button" class="btn dropdown-toggle btn-secondary btn-sm"><span>kieli-sisalto:
      <span class="valittukieli">fi</span> <span aria-hidden="true" class="icon material-icons">expand_more</span></span></button>
  <ul role="menu" tabindex="-1" class="dropdown-menu dropdown-menu-right">
    <li role="presentation"><a role="menuitem" href="#" target="_self" tabindex="-1" aria-disabled="true" class="dropdown-item disabled">fi</a></li>
    <li role="presentation"><a role="menuitem" href="#" target="_self" class="dropdown-item">sv</a></li>
    <li role="presentation"><a role="menuitem" href="#" target="_self" class="dropdown-item">en</a></li>
  </ul>
</div>
`;
